src/uuid.ts

```diff
@@ -1,57 +1,219 @@
+  toHexString(): string {
+    if (UUID.cacheHexString && this.__id) {
+      return this.__id;
+    }
 
-  // Parse ########-....-....-....-............
-  arr[0] = (v = parseInt(uuid.slice(0, 8), 16)) >>> 24;
-  arr[1] = (v >>> 16) & 0xff;
-  arr[2] = (v >>> 8) & 0xff;
-  arr[3] = v & 0xff;
+    const uuidHexString = [
+      this.id.toString('hex', 0, 4),
+      this.id.toString('hex', 4, 6),
+      this.id.toString('hex', 6, 8),
+      this.id.toString('hex', 8, 10),
+      this.id.toString('hex', 10, 16)
+    ].join('-');
 
-  // Parse ........-####-....-....-............
-  arr[4] = (v = parseInt(uuid.slice(9, 13), 16)) >>> 8;
-  arr[5] = v & 0xff;
+    if (UUID.cacheHexString) {
+      this.__id = uuidHexString;
+    }
 
-  // Parse ........-....-####-....-............
-  arr[6] = (v = parseInt(uuid.slice(14, 18), 16)) >>> 8;
-  arr[7] = v & 0xff;
+    return uuidHexString;
+  }
 
-  // Parse ........-....-....-####-............
-  arr[8] = (v = parseInt(uuid.slice(19, 23), 16)) >>> 8;
-  arr[9] = v & 0xff;
+  /**
+   * Converts the id into a 36 character (dashes included) hex string, unless a encoding is specified.
+   * @internal
+   */
+  toString(encoding?: string): string {
+    return encoding ? this.id.toString(encoding) : this.toHexString();
+  }
+
+  /**
+   * Converts the id into its JSON string representation. A 36 character (dashes included) hex string in the format: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+   * @internal
+   */
+  toJSON(): string {
+    return this.toHexString();
+  }
 
-  // Parse ........-....-....-....-############
-  // (Use "/" to avoid 32-bit truncation when bit-shifting high-order bytes)
-  arr[10] = ((v = parseInt(uuid.slice(24, 36), 16)) / 0x10000000000) & 0xff;
-  arr[11] = (v / 0x100000000) & 0xff;
-  arr[12] = (v >>> 24) & 0xff;
-  arr[13] = (v >>> 16) & 0xff;
-  arr[14] = (v >>> 8) & 0xff;
-  arr[15] = v & 0xff;
+  /**
+   * Compares the equality of this UUID with `otherID`.
+   *
+   * @param otherId - UUID instance to compare against.
+   */
+  equals(otherId: string | Buffer | UUID): boolean {
+    if (!otherId) {
+      return false;
+    }
 
-  return arr;
+    if (otherId instanceof UUID) {
+      return otherId.id.equals(this.id);
+    }
+
+    try {
+      return new UUID(otherId).id.equals(this.id);
+    } catch (_) {
+      return false;
+    }
+  }
+
+  /**
+   * Checks if a value is a valid bson UUID
+   * @param input - UUID, string or Buffer to validate.
+   */
+  static isValid(input: string | Buffer | UUID): boolean {
+    if (!input) {
+      return false;
+    }
+
+    if (input instanceof UUID) {
+      return true;
+    }
+
+    if (typeof input === 'string') {
+      return uuidHexStringValidateV4(input);
+    }
+
+    if (Buffer.isBuffer(input) && input.length === BYTE_LENGTH) {
+      // TODO: Investigate a faster check, than converting to a string and letting uuid handle the rest?
+      return uuidHexStringValidateV4(bufferToUuidHexString(input));
```

Comment:
should this check be guarded. I guess isValid should not throw.

Reply:
Yeah, something has to happen here, and I've probably gone a bit too far with the validation (if you compare to the existing one in [ObjectId](https://github.com/steffenagger/js-bson/blob/master/src/objectid.ts#L316-L318)).
This is one of the areas I would like the node.js driver teams input/knowledge on... this part could very well end up just returning `true` 🤷‍♂️ (if not, I'll guard it).

Reply:
Reworked.